processing/src/main/java/org/apache/druid/frame/field/StringFieldReader.java

```diff
@@ -0,0 +1,276 @@
+  {
+    return new Selector(memory, fieldPointer, null, asArray);
+  }
+
+  @Override
+  public DimensionSelector makeDimensionSelector(
+      Memory memory,
+      ReadableFieldPointer fieldPointer,
+      @Nullable ExtractionFn extractionFn
+  )
+  {
+    if (asArray) {
+      throw new ISE("Cannot call makeDimensionSelector on field of type [%s]", ColumnType.STRING_ARRAY);
+    }
+
+    return new Selector(memory, fieldPointer, extractionFn, false);
+  }
+
+  @Override
+  public boolean isComparable()
+  {
+    return true;
+  }
+
+  private static class Selector implements DimensionSelector
+  {
+    private final Memory memory;
+    private final ReadableFieldPointer fieldPointer;
+    @Nullable
+    private final ExtractionFn extractionFn;
+    private final boolean asArray;
+
+    private long currentFieldPosition = -1;
+    private final RangeIndexedInts indexedInts = new RangeIndexedInts();
+    private final List<ByteBuffer> currentUtf8Strings = new ArrayList<>();
+
+    private Selector(
+        final Memory memory,
+        final ReadableFieldPointer fieldPointer,
+        @Nullable final ExtractionFn extractionFn,
+        final boolean asArray
+    )
+    {
+      this.memory = memory;
+      this.fieldPointer = fieldPointer;
+      this.extractionFn = extractionFn;
+      this.asArray = asArray;
+    }
+
+    @Nullable
+    @Override
+    public Object getObject()
+    {
+      final List<ByteBuffer> currentStrings = computeCurrentUtf8Strings();
+      final int size = currentStrings.size();
+
+      if (size == 0) {
+        return asArray ? Collections.emptyList() : null;
+      } else if (size == 1) {
+        return asArray ? Collections.singletonList(lookupName(0)) : lookupName(0);
+      } else {
+        final List<String> strings = new ArrayList<>(size);
+        for (int i = 0; i < size; i++) {
+          strings.add(lookupName(i));
+        }
+        return strings;
+      }
+    }
+
+    @Override
+    public IndexedInts getRow()
+    {
+      indexedInts.setSize(computeCurrentUtf8Strings().size());
+      return indexedInts;
+    }
+
+    @Nullable
+    @Override
+    public String lookupName(int id)
+    {
+      final ByteBuffer byteBuffer = computeCurrentUtf8Strings().get(id);
+      final String s = byteBuffer != null ? StringUtils.fromUtf8(byteBuffer.duplicate()) : null;
+      return extractionFn == null ? s : extractionFn.apply(s);
+    }
+
+    @Override
+    public boolean supportsLookupNameUtf8()
+    {
+      return extractionFn == null;
+    }
+
+    @Nullable
+    @Override
+    public ByteBuffer lookupNameUtf8(int id)
+    {
+      if (extractionFn != null) {
+        throw new ISE("Cannot use lookupNameUtf8 on this selector");
+      }
+
+      return computeCurrentUtf8Strings().get(id);
+    }
+
+    @Override
+    public int getValueCardinality()
+    {
+      return CARDINALITY_UNKNOWN;
+    }
```

Comment:
Some of these attributes seem related more to the *type* of the data than the actual *data*. Should there be a reader which can deliver strings, and can provide an object with the type info? That is, separate the representation into data and metadata?

Reply:
This is implementing DimensionSelector, an interface we've had since time immemorial. It's used for reading string and multi-value string columns. I think @clintropolis has been working on teasing some of these things apart, as part of the effort to enable dictionary coding and indexing for things that aren't strings. So he may be able to comment more on future plans for this interface. In this patch, I'm taking it as it is.

Reply:
I got a bit side-tracked with some other stuff and haven't got back to `DimensionSelector` yet, but still thinking about it in the back of my mind. I mainly want to do 2 things to it, 1) decouple the dictionary encoded selector part from strings so we can make more easily make dictionary encoded selectors for other column types and use them in places, and 2) split out the 'use it like an array' usages from the dictionary encoded uses, since its the reason we have to decorate these selectors with stuff like the cardinality and methods like 'nameLookupPossibleInAdvance', so that callers can check to see what type of selector they got so they know what they can do with it. I would agree the selector itself might not be the best place for this stuff to live, so will keep this in mind.